sparse/impl/KokkosSparse_spgemm_cuSPARSE_impl.hpp

```diff
@@ -172,8 +174,89 @@ void cuSPARSE_symbolic(KernelHandle *handle, typename KernelHandle::nnz_lno_t m,
+  handle->create_cusparse_spgemm_handle(transposeA, transposeB);
+  typename KernelHandle::cuSparseSpgemmHandleType *h =
+      handle->get_cusparse_spgemm_handle();
+
+  // Follow
+  // https://github.com/NVIDIA/CUDALibrarySamples/blob/master/cuSPARSE/spgemm
+
+  const auto alpha = Kokkos::ArithTraits<scalar_type>::one();
+  const auto beta  = Kokkos::ArithTraits<scalar_type>::zero();
+
+  // In non-reuse interface, forced to give A,B dummy values to
+  // cusparseSpGEMM_compute. And it actually reads them, so they must be
+  // allocated and of the correct type. This compute will be called again in
+  // numeric with the real values.
+  //
+  // The dummy values can be uninitialized. cusparseSpGEMM_compute does
+  // not remove numerical zeros from the sparsity pattern.
+  void *dummyValues;
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc(
+      &dummyValues,
+      sizeof(scalar_type) * std::max(entriesA.extent(0), entriesB.extent(0))));
+
+  KOKKOS_CUSPARSE_SAFE_CALL(cusparseCreateCsr(
+      &h->descr_A, m, n, entriesA.extent(0), (void *)row_mapA.data(),
+      (void *)entriesA.data(), dummyValues, CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_32I, CUSPARSE_INDEX_BASE_ZERO, h->scalarType));
+
+  KOKKOS_CUSPARSE_SAFE_CALL(cusparseCreateCsr(
+      &h->descr_B, n, k, entriesB.extent(0), (void *)row_mapB.data(),
+      (void *)entriesB.data(), dummyValues, CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_32I, CUSPARSE_INDEX_BASE_ZERO, h->scalarType));
+
+  KOKKOS_CUSPARSE_SAFE_CALL(cusparseCreateCsr(
+      &h->descr_C, m, k, 0, NULL, NULL, NULL, CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_32I, CUSPARSE_INDEX_BASE_ZERO, h->scalarType));
+
+  //----------------------------------------------------------------------
+  // query workEstimation buffer size, allocate, then call again with buffer.
+  KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_workEstimation(
+      h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B, &beta,
+      h->descr_C, h->scalarType, h->alg, h->spgemmDescr, &h->bufferSize3,
+      NULL));
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc((void **)&h->buffer3, h->bufferSize3));
+  KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_workEstimation(
+      h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B, &beta,
+      h->descr_C, h->scalarType, h->alg, h->spgemmDescr, &h->bufferSize3,
+      h->buffer3));
+
+  //----------------------------------------------------------------------
+  // query compute buffer size, allocate, then call again with buffer.
+
+  KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_compute(
+      h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B, &beta,
+      h->descr_C, h->scalarType, CUSPARSE_SPGEMM_DEFAULT, h->spgemmDescr,
+      &h->bufferSize4, NULL));
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc((void **)&h->buffer4, h->bufferSize4));
+  KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_compute(
+      h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B, &beta,
```

Comment:
Swap with `alpha` and `beta` user passes in?

Reply:
@e10harvey Our SpGEMM interface actually doesn't take alpha or beta, and I wasn't planning on adding them. ``beta != 0`` would mean doing a fused SpGEMM+SpAdd (``C := AB + beta*C``), which is possible but would be a lot of work and would mean a whole new symbolic and numeric implementation.

Alpha (``C := alpha * AB``) would be a lot easier to support, but there's also not a KokkosKernels use case for it that I know of. If this was needed, the user could either apply the scaling themselves when C is used, or just do a ``KokkosBlas::scal`` on C's values.